File: backend/app/schemas/tasks.py
from datetime import datetime
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator


class TaskCreate(BaseModel):
    """Create a task from an inline requirement or an existing requirement."""

    requirement_id: Optional[int] = Field(default=None, ge=1)
    title: Optional[str] = Field(default=None, max_length=120)
    description: Optional[str] = Field(default=None, max_length=20000)
    # None means “not provided”; [] is an explicit request to clear tags on
    # an existing requirement.
    focus_tags: Optional[List[str]] = Field(default=None, max_length=30)
    model_id: Optional[int] = Field(default=None, ge=1)
    prompt_template_id: Optional[int] = Field(default=None, ge=1)
    auto_review: bool = False
    run_generate: bool = False
    # Compatibility clients may omit this; the API explicitly resolves the
    # default space while the new frontend always sends it.
    wiki_space_id: Optional[int] = Field(default=None, ge=1)
    generation_granularity: Literal["compact", "standard", "detailed"] = "standard"
    test_dimensions: List[str] = Field(default_factory=lambda: ["positive", "negative", "boundary"])
    # ``dimensions`` is accepted as a short-lived compatibility spelling for
    # clients built against the design draft.  The response always exposes
    # the canonical ``test_dimensions`` field.
    dimensions: Optional[List[str]] = None
    reference_case_ids: List[int] = Field(default_factory=list, max_length=10)
    reference_text: str = Field(default="", max_length=16000)

    @field_validator("focus_tags")
    @classmethod
    def validate_focus_tags(cls, value: Optional[List[str]]) -> Optional[List[str]]:
        if value is None:
            return None
        if any(not isinstance(item, str) or len(item.strip()) > 80 for item in value):
            raise ValueError("Each focus tag must be at most 80 characters")
        return [item.strip() for item in value]

    @field_validator("reference_case_ids")
    @classmethod
    def validate_reference_case_ids(cls, value: List[int]) -> List[int]:
        if any(item <= 0 for item in value):
            raise ValueError("reference_case_ids must contain positive ids")
        return value


class ReviewResultOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    task_id: int
    draft_id: int
    score: int
    verdict: str
    payload: Dict[str, Any] = Field(default_factory=dict)
    created_at: datetime


class TaskOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    requirement_id: int
    wiki_space_id: int
    wiki_space_name: str = ""
    status: str
    model_id: Optional[int] = None
    prompt_template_id: Optional[int] = None
    error_message: Optional[str] = None
    title: Optional[str] = None
    description: Optional[str] = None
    focus_tags: List[str] = Field(default_factory=list)
    citation_count: int = 0
    latest_draft_snippet: Optional[str] = None
    latest_draft_version: Optional[int] = None
    latest_review: Optional[ReviewResultOut] = None
    finalized_draft_id: Optional[int] = None
    finalized_at: Optional[datetime] = None
    imported_case_ids: List[int] = Field(default_factory=list)
    imported_case_count: int = 0
    generation_granularity: str = "standard"
    test_dimensions: List[str] = Field(default_factory=list)
    reference_case_count: int = 0
    test_point_count: int = 0
    created_at: datetime
    updated_at: datetime


class TaskCitationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    path: str
    score: float = 0.0
    snippet: str = ""
    wiki_page_id: Optional[int] = None
    citation_type: str = "wiki"
    source_chunk_id: Optional[int] = None
    content_excerpt: str = ""
    clause_ids: List[str] = Field(default_factory=list)
    anchor_clause: Optional[str] = None
    available: bool = True
    legacy: bool = False
    legacy_reason: Optional[str] = None


class RetrievalCheckpointOut(BaseModel):
    id: int
    task_id: int
    attempt: int
    version: int
    status: str
    auto_review: bool = False
    query: str
    candidate_citations: List[TaskCitationOut] = Field(default_factory=list)
    selected_citation_ids: List[int] = Field(default_factory=list)
    supplemental_text: str = ""
    idempotency_key: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class RetrievalCheckpointConfirm(BaseModel):
    selected_citation_ids: List[int] = Field(default_factory=list)
    supplemental_text: str = Field(default="", max_length=10000)
    expected_version: int = Field(ge=1)
    idempotency_key: str = Field(min_length=1, max_length=200)


class TaskReferenceCaseOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    task_id: int
    source_case_id: Optional[int] = None
    source_case_key: Optional[str] = None
    title_snapshot: str
    content_md_snapshot: str
    content_hash: str
    source: str
    created_at: datetime


class TestPointInput(BaseModel):
    id: Optional[int] = Field(default=None, ge=1)
    stable_key: str = Field(min_length=1, max_length=80)
    title: str = Field(min_length=1, max_length=240)
    verification_goal: str = Field(default="", max_length=1000)
    dimension: str = Field(default="positive", min_length=1, max_length=40)
    priority: Literal["P0", "P1", "P2"] = "P1"
    sort_order: int = Field(default=0, ge=0, le=10000)
    is_selected: bool = True
    is_excluded: bool = False
    citation_ids: List[int] = Field(default_factory=list, max_length=50)


class TestPointOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    task_id: int
    checkpoint_id: int
    stable_key: str
    title: str
    verification_goal: str
    dimension: str
    priority: str
    sort_order: int
    is_selected: bool
    is_excluded: bool
    citation_ids: List[int] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class TestPointCheckpointOut(BaseModel):
    id: int
    task_id: int
    retrieval_checkpoint_id: Optional[int] = None
    attempt: int
    version: int
    status: str
    points: List[TestPointOut] = Field(default_factory=list)
    idempotency_key: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class TestPointEditRequest(BaseModel):
    points: List[TestPointInput] = Field(max_length=200)
    expected_version: int = Field(ge=1)


class TestPointConfirmRequest(TestPointEditRequest):
    idempotency_key: str = Field(min_length=1, max_length=200)


class RequirementOptimizeRequest(BaseModel):
    title: str = Field(min_length=1, max_length=120)
    description: str = Field(min_length=1, max_length=20000)
    focus_tags: List[str] = Field(default_factory=list, max_length=30)
    model_id: Optional[int] = Field(default=None, ge=1)


class RequirementOptimizeOut(BaseModel):
    title: str
    description: str
    questions: List[str] = Field(default_factory=list)
    prompt_type: str = "requirement_optimize"


class CaseDraftOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    task_id: int
    version: int
    content_md: str
    prompt_version_ref: Optional[str] = None
    created_at: datetime


class FinalizeTaskBody(BaseModel):
    """Optional exact draft selection for finalization.

    The body is optional at the route level for compatibility with the legacy
    empty POST; when omitted the latest draft is selected.
    """

    draft_id: Optional[int] = Field(default=None, ge=1)


class TestCaseOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    requirement_id: int
    case_key: str
    # ``source_case_key`` is retained as a first-class response field because
    # import identity must remain visible even after a manual edit.
    source_case_key: Optional[str] = None
    title: str = ""
    content_md: str
    # A small compatibility alias for clients that call the editable body
    # simply ``content``.
    content: Optional[str] = None
    status: str
    priority: str = "P1"
    revision: int
    source_task_id: Optional[int] = None
    source_draft_id: Optional[int] = None
    source_draft_version: Optional[int] = None
    archived_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class TestCaseUpdate(BaseModel):
    title: Optional[str] = None
    content_md: Optional[str] = None
    content: Optional[str] = None
    expected_revision: Optional[int] = Field(default=None, ge=1)
    revision: Optional[int] = Field(default=None, ge=1)
    expected_updated_at: Optional[datetime] = None
    reason: Optional[str] = None
    priority: Optional[Literal["P0", "P1", "P2"]] = None


class TestCaseOperationLogOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    test_case_id: int
    operation: str
    changed_fields: List[str] = Field(default_factory=list)
    before_hash: Optional[str] = None
    after_hash: Optional[str] = None
    before_length: Optional[int] = None
    after_length: Optional[int] = None
    added_lines: int = 0
    deleted_lines: int = 0
    title_changed: bool = False
    diff_summary: str = ""
    reason: Optional[str] = None
    operator: Optional[str] = None
    source_task_id: Optional[int] = None
    source_draft_id: Optional[int] = None
    source_case_key: Optional[str] = None
    created_at: datetime


class TestCaseCreate(BaseModel):
    requirement_id: int = Field(ge=1)
    case_key: str = Field(min_length=1)
    title: Optional[str] = None
    content_md: Optional[str] = None
    content: Optional[str] = None
    priority: Literal["P0", "P1", "P2"] = "P1"


class TaskEventOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    task_id: int
    step: str
    message: str
    detail_json: Optional[str] = None
    created_at: datetime
    wiki_space_id: Optional[int] = None
    wiki_space_name: str = ""


class ApplyPromptBody(BaseModel):
    revision_id: int
    mode: Literal["global", "task_temp"]
    content: Optional[str] = Field(default=None, min_length=1)


class TaskModelUpdate(BaseModel):
    """Select the model used by the next generation attempt."""

    model_id: Optional[int] = Field(default=None, ge=1)


class PromptRevisionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    task_id: int
    base_prompt_id: Optional[int] = None
    new_content: str
    status: str
    created_at: datetime


class CoverageTestPointOut(BaseModel):
    stable_key: str
    title: str
    priority: str
    dimension: str
    selected: bool
    excluded: bool
    covered: bool
    case_ids: List[int] = Field(default_factory=list)
    citation_ids: List[int] = Field(default_factory=list)


class CoverageCitationOut(BaseModel):
    citation_id: int
    title: str
    path: str
    test_point_keys: List[str] = Field(default_factory=list)
    case_ids: List[int] = Field(default_factory=list)
    used: bool = False


class CoverageSummaryOut(BaseModel):
    task_id: int
    total_test_points: int
    selected_test_points: int
    covered_test_points: int
    uncovered_test_points: int
    coverage_percent: float
    points: List[CoverageTestPointOut] = Field(default_factory=list)
    citations: List[CoverageCitationOut] = Field(default_factory=list)

File: frontend/src/api/tasks.ts
import { api } from './client'

export interface ReviewResult {
  id: number
  task_id: number
  draft_id: number
  score: number
  verdict: string
  payload: {
    score?: number
    verdict?: string
    issues?: string[]
    missing_scenarios?: string[]
    prompt_improvement_hints?: string[]
    ready_for_final?: boolean
    raw?: string
    [key: string]: unknown
  }
  created_at: string
}

export interface TaskItem {
  id: number
  requirement_id: number
  wiki_space_id: number
  wiki_space_name: string
  status: string
  auto_review?: boolean
  model_id: number | null
  prompt_template_id: number | null
  error_message: string | null
  title: string | null
  description: string | null
  focus_tags: string[]
  citation_count: number
  latest_draft_snippet: string | null
  latest_draft_version: number | null
  latest_review: ReviewResult | null
  finalized_draft_id?: number | null
  finalized_at?: string | null
  imported_case_ids?: number[]
  imported_case_count?: number
  generation_granularity: 'compact' | 'standard' | 'detailed' | string
  test_dimensions: string[]
  reference_case_count: number
  test_point_count: number
  created_at: string
  updated_at: string
}

export interface TaskCreate {
  requirement_id?: number | null
  title: string
  description: string
  focus_tags?: string[]
  model_id?: number | null
  prompt_template_id?: number | null
  auto_review?: boolean
  run_generate?: boolean
  wiki_space_id?: number
  generation_granularity?: 'compact' | 'standard' | 'detailed'
  test_dimensions?: string[]
  reference_case_ids?: number[]
  reference_text?: string
}

export interface CaseDraft {
  id: number
  task_id: number
  version: number
  content_md: string
  prompt_version_ref: string | null
  created_at: string
}

export interface TaskEvent {
  id: number
  task_id: number
  step: string
  message: string
  detail_json: string | null
  created_at: string
  wiki_space_id?: number | null
  wiki_space_name?: string
}

export interface TaskStreamPayload {
  status?: string
  message?: string
  text?: string
  delta?: string
  sequence?: number
  terminal?: 'completed' | 'failed' | null
  attempt?: number
}

export interface PromptRevision {
  id: number
  task_id: number
  base_prompt_id: number | null
  new_content: string
  status: string
  created_at: string
}

export interface TaskCitation {
  id: number
  title: string
  path: string
  score: number
  snippet: string
  wiki_page_id: number | null
  citation_type?: string
  source_chunk_id?: number | null
  content_excerpt?: string
  clause_ids?: string[]
  anchor_clause?: string | null
  available?: boolean
  legacy?: boolean
  legacy_reason?: string | null
}

export interface RetrievalCheckpoint {
  id: number
  task_id: number
  attempt: number
  version: number
  status: string
  query: string
  candidate_citations: TaskCitation[]
  selected_citation_ids: number[]
  supplemental_text: string
  idempotency_key: string | null
  created_at: string
  updated_at: string
}

export interface TaskReferenceCase {
  id: number
  task_id: number
  source_case_id: number | null
  source_case_key: string | null
  title_snapshot: string
  content_md_snapshot: string
  content_hash: string
  source: 'case_library' | 'manual' | string
  created_at: string
}

export interface TestPointItem {
  id: number
  task_id: number
  checkpoint_id: number
  stable_key: string
  title: string
  verification_goal: string
  dimension: string
  priority: 'P0' | 'P1' | 'P2' | string
  sort_order: number
  is_selected: boolean
  is_excluded: boolean
  citation_ids: number[]
  created_at: string
  updated_at: string
}

export interface TestPointCheckpoint {
  id: number
  task_id: number
  retrieval_checkpoint_id: number | null
  attempt: number
  version: number
  status: string
  points: TestPointItem[]
  idempotency_key: string | null
  created_at: string
  updated_at: string
}

export interface CoveragePoint {
  stable_key: string
  title: string
  priority: string
  dimension: string
  selected: boolean
  excluded: boolean
  covered: boolean
  case_ids: number[]
  citation_ids: number[]
}

export interface CoverageCitation {
  citation_id: number
  title: string
  path: string
  test_point_keys: string[]
  case_ids: number[]
  used: boolean
}

export interface CoverageSummary {
  task_id: number
  total_test_points: number
  selected_test_points: number
  covered_test_points: number
  uncovered_test_points: number
  coverage_percent: number
  points: CoveragePoint[]
  citations: CoverageCitation[]
}

export type ApplyPromptMode = 'global' | 'task_temp'

export const IN_PROGRESS_STATUSES = new Set([
  'retrieving',
  'generating',
  'reviewing',
  'optimizing',
  'regenerating',
  'generating_test_points',
])

export function shouldPollTaskStatus(status: string | null | undefined): boolean {
  return Boolean(status && IN_PROGRESS_STATUSES.has(status))
}

export function listTasks() {
  return api<TaskItem[]>('/api/tasks')
}

export function getTask(id: number) {
  return api<TaskItem>(`/api/tasks/${id}`)
}

export function taskStreamUrl(id: number) {
  return `/api/tasks/${id}/stream`
}

export function deleteTask(id: number) {
  return api<{ ok: boolean; id: number }>(`/api/tasks/${id}`, { method: 'DELETE' })
}

export function updateTaskModel(id: number, modelId: number | null) {
  return api<TaskItem>(`/api/tasks/${id}/model`, {
    method: 'PATCH',
    body: JSON.stringify({ model_id: modelId }),
  })
}

export function createTask(body: TaskCreate) {
  return api<TaskItem>('/api/tasks', {
    method: 'POST',
    body: JSON.stringify(body),
  })
}

export function generateTask(id: number, opts?: { auto_review?: boolean }) {
  const q = opts?.auto_review ? '?auto_review=true' : ''
  return api<TaskItem>(`/api/tasks/${id}/generate${q}`, { method: 'POST' })
}

export function reviewTask(id: number) {
  return api<TaskItem>(`/api/tasks/${id}/review`, { method: 'POST' })
}

export function optimizePromptTask(id: number) {
  return api<TaskItem>(`/api/tasks/${id}/optimize-prompt`, { method: 'POST' })
}

export function regenerateTask(id: number) {
  return api<TaskItem>(`/api/tasks/${id}/regenerate`, { method: 'POST' })
}

export function finalizeTask(id: number, draftId?: number | null) {
  return api<TaskItem>(`/api/tasks/${id}/finalize`, {
    method: 'POST',
    ...(draftId == null ? {} : { body: JSON.stringify({ draft_id: draftId }) }),
  })
}

export function applyPrompt(
  id: number,
  body: { revision_id: number; mode: ApplyPromptMode; content?: string },
) {
  return api<TaskItem>(`/api/tasks/${id}/apply-prompt`, {
    method: 'POST',
    body: JSON.stringify(body),
  })
}

export function listDrafts(id: number) {
  return api<CaseDraft[]>(`/api/tasks/${id}/drafts`)
}

export function listEvents(id: number) {
  return api<TaskEvent[]>(`/api/tasks/${id}/events`)
}

export function listReviews(id: number) {
  return api<ReviewResult[]>(`/api/tasks/${id}/reviews`)
}

export function listRevisions(id: number) {
  return api<PromptRevision[]>(`/api/tasks/${id}/revisions`)
}

export function listCitations(id: number) {
  return api<TaskCitation[]>(`/api/tasks/${id}/citations`)
}

export function getRetrievalCheckpoint(id: number) {
  return api<RetrievalCheckpoint>(`/api/tasks/${id}/retrieval-checkpoint`)
}

export function confirmRetrievalCheckpoint(id: number, body: { selected_citation_ids: number[]; supplemental_text: string; expected_version: number; idempotency_key: string }) {
  return api<TaskItem>(`/api/tasks/${id}/retrieval-checkpoint/confirm`, {
    method: 'POST',
    body: JSON.stringify(body),
  })
}

export function optimizeRequirement(body: {
  title: string
  description: string
  focus_tags?: string[]
  model_id?: number | null
}) {
  return api<{
    title: string
    description: string
    questions: string[]
    prompt_type: string
  }>('/api/tasks/requirement-optimize', {
    method: 'POST',
    body: JSON.stringify(body),
  })
}

export function getTestPointCheckpoint(id: number) {
  return api<TestPointCheckpoint>(`/api/tasks/${id}/test-points`)
}

export function editTestPoints(id: number, body: {
  points: Array<Partial<TestPointItem> & {
    stable_key: string
    title: string
    verification_goal: string
    dimension: string
    priority: string
    sort_order: number
    is_selected: boolean
    is_excluded: boolean
    citation_ids: number[]
  }>
  expected_version: number
}) {
  return api<TestPointCheckpoint>(`/api/tasks/${id}/test-points`, {
    method: 'PUT',
    body: JSON.stringify(body),
  })
}

export function confirmTestPoints(id: number, body: {
  points: Array<{
    stable_key: string
    title: string
    verification_goal: string
    dimension: string
    priority: string
    sort_order: number
    is_selected: boolean
    is_excluded: boolean
    citation_ids: number[]
  }>
  expected_version: number
  idempotency_key: string
}) {
  return api<TaskItem>(`/api/tasks/${id}/test-points/confirm`, {
    method: 'POST',
    body: JSON.stringify(body),
  })
}

export function listTaskReferences(id: number) {
  return api<TaskReferenceCase[]>(`/api/tasks/${id}/references`)
}

export function getTaskCoverage(id: number) {
  return api<CoverageSummary>(`/api/tasks/${id}/coverage`)
}

export function statusLabel(status: string): string {
  const map: Record<string, string> = {
    draft: '草稿',
    retrieving: '检索中',
    awaiting_confirmation: '等待确认',
    generating_test_points: '生成测试点中',
    awaiting_test_point_confirmation: '等待测试点确认',
    generating: '生成中',
    generated: '已生成',
    reviewing: '评审中',
    reviewed: '已评审',
    optimizing: '优化中',
    regenerating: '再生成中',
    finalized: '已终版',
    failed: '失败',
  }
  return map[status] || status
}

export function statusTagType(status: string): '' | 'success' | 'warning' | 'info' | 'danger' {
  switch (status) {
    case 'finalized':
      return 'success'
    case 'generated':
    case 'reviewed':
      return 'info'
    case 'retrieving':
    case 'awaiting_confirmation':
    case 'generating_test_points':
    case 'awaiting_test_point_confirmation':
    case 'generating':
    case 'reviewing':
    case 'optimizing':
    case 'regenerating':
      return 'warning'
    case 'failed':
      return 'danger'
    default:
      return ''
  }
}
